# Deep Learning - Question 2
 
#### Change the data source to Breast Cancerdataset * available in the source folder https://umkc.box.com/s/3cvfiwc81lhgygc67deyeqs8m858lld0In  and make required changes.

In [1]:
#Loading the Libraries

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers.core import Dense
from sklearn.preprocessing import LabelEncoder

In [2]:
#Importing the data
df = pd.read_csv('Breas Cancer.csv')

In [3]:
#Printing the first five rows
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [4]:
#Finding the overal info about the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

##### Observations:
1. There are 32 columns
2. Only one column is categorical (diagnosis)

In [5]:
#Finding the unique value or diagnosis column since it is only categorical value in data
df['diagnosis'].unique()

array(['M', 'B'], dtype=object)

In [6]:
#Finding any values in Unnamed: 32 column
df['Unnamed: 32'].unique()

array([nan])

#### As we could see there are no values in Unnamed: 32 columns, so we can skip this feature

In [7]:
df = df.drop(['Unnamed: 32'],axis = 1)

##### Converting the categorical data to int

In [8]:
#Converting the categorical data to int
convert = {"diagnosis": {"M": 0, "B": 1}}
cancer_data = df.replace(convert)
cancer_data.head()

id  diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302          0        17.99         10.38          122.80     1001.0   
1    842517          0        20.57         17.77          132.90     1326.0   
2  84300903          0        19.69         21.25          130.00     1203.0   
3  84348301          0        11.42         20.38           77.58      386.1   
4  84358402          0        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

In [9]:
#Copying the remaining features other than diagnosis to x
x = cancer_data.iloc[:,2:31]

In [10]:
#Copying the diagnosis feature(Predictable feature)
y = cancer_data['diagnosis']

In [11]:
#splitting the data into training and testing categories
X_train, X_test, Y_train, Y_test = train_test_split(x, y,test_size=0.25, random_state=87)

In [14]:
#Building the sequential model
np.random.seed(155)
my_first_nn = Sequential() # create model
my_first_nn.add(Dense(20, input_dim=29, activation='relu')) # hidden layer
my_first_nn.add(Dense(1, activation='sigmoid')) # output layer

#Performing compilation
my_first_nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['acc'])
my_first_nn_fitted = my_first_nn.fit(X_train, Y_train, epochs=100,initial_epoch=0)

#Printing Summary and Accuracy
print(my_first_nn.summary())
print(my_first_nn.evaluate(X_test, Y_test))

Epoch 1/100
14/14 [==============================] - 0s 2ms/step - loss: 19.6513 - acc: 0.6197
Epoch 2/100
14/14 [==============================] - 0s 2ms/step - loss: 7.1139 - acc: 0.5305
Epoch 3/100
14/14 [==============================] - 0s 2ms/step - loss: 3.0576 - acc: 0.3638
Epoch 4/100
14/14 [==============================] - 0s 2ms/step - loss: 2.2402 - acc: 0.2817
Epoch 5/100
14/14 [==============================] - 0s 3ms/step - loss: 1.8448 - acc: 0.2934
Epoch 6/100
14/14 [==============================] - 0s 3ms/step - loss: 1.5579 - acc: 0.2981
Epoch 7/100
14/14 [==============================] - 0s 3ms/step - loss: 1.3252 - acc: 0.3333
Epoch 8/100
14/14 [==============================] - 0s 3ms/step - loss: 1.1400 - acc: 0.3592
Epoch 9/100
14/14 [==============================] - 0s 3ms/step - loss: 0.9915 - acc: 0.4108
Epoch 10/100
14/14 [==============================] - 0s 3ms/step - loss: 0.8585 - acc: 0.4671
Epoch 11/100
14/14 [==============================] - 0s 2

14/14 [==============================] - 0s 2ms/step - loss: 0.1963 - acc: 0.9249
Epoch 88/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2005 - acc: 0.9296
Epoch 89/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1990 - acc: 0.9249
Epoch 90/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1898 - acc: 0.9272
Epoch 91/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2004 - acc: 0.9225
Epoch 92/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2059 - acc: 0.9155
Epoch 93/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1893 - acc: 0.9225
Epoch 94/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1776 - acc: 0.9366
Epoch 95/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2087 - acc: 0.9202
Epoch 96/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1923 - acc: 0.9272
Epoch 97/100
14/14 [==============================] - 0s 2ms/st

#### Observations :
1. Loss     = 26%
2. Accuracy = 89%

### Adding more dense layers 

In [16]:
#Building the sequential model
np.random.seed(155)
my_first_nn = Sequential() # create model
my_first_nn.add(Dense(20, input_dim=29, activation='relu')) # hidden layer
my_first_nn.add(Dense(40, activation='relu')) #adding more hidden layers
my_first_nn.add(Dense(1, activation='sigmoid')) # output layer

#Performing compilation
my_first_nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['acc'])
my_first_nn_fitted = my_first_nn.fit(X_train, Y_train, epochs=100,initial_epoch=0)

#Printing Summary and Accuracy
print(my_first_nn.summary())
print(my_first_nn.evaluate(X_test, Y_test))

Epoch 1/100
14/14 [==============================] - 0s 2ms/step - loss: 13.2998 - acc: 0.4178
Epoch 2/100
14/14 [==============================] - 0s 2ms/step - loss: 3.4443 - acc: 0.4977
Epoch 3/100
14/14 [==============================] - 0s 2ms/step - loss: 1.0246 - acc: 0.6737
Epoch 4/100
14/14 [==============================] - 0s 3ms/step - loss: 0.5372 - acc: 0.7958
Epoch 5/100
14/14 [==============================] - 0s 3ms/step - loss: 0.4172 - acc: 0.8474
Epoch 6/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2607 - acc: 0.9085
Epoch 7/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2292 - acc: 0.9225
Epoch 8/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2250 - acc: 0.9249
Epoch 9/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2382 - acc: 0.8991
Epoch 10/100
14/14 [==============================] - 0s 2ms/step - loss: 0.2204 - acc: 0.9225
Epoch 11/100
14/14 [==============================] - 0s 2

14/14 [==============================] - 0s 3ms/step - loss: 0.1659 - acc: 0.9366
Epoch 88/100
14/14 [==============================] - 0s 3ms/step - loss: 0.1896 - acc: 0.9343
Epoch 89/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1496 - acc: 0.9343
Epoch 90/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1776 - acc: 0.9343
Epoch 91/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1932 - acc: 0.9319
Epoch 92/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1878 - acc: 0.9272
Epoch 93/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1543 - acc: 0.9390
Epoch 94/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1743 - acc: 0.9366
Epoch 95/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1528 - acc: 0.9460
Epoch 96/100
14/14 [==============================] - 0s 2ms/step - loss: 0.1500 - acc: 0.9343
Epoch 97/100
14/14 [==============================] - 0s 2ms/st

#### Observations after adding extra hidden layer:
1. Loss     = 24%
2. Accuracy = 90%